File: seungtae/23_09/programmers/77486.java
import java.util.*;

public class Solution {
	public int[] solution(String[] enroll, String[] referral, String[] seller, int[] amount) {

		Map<String, Integer> revenue = new HashMap<>();
		Map<String, String> parent = new HashMap<>();

		// 수익 딕셔너리 초기화
		for (String s: enroll) {
			revenue.put(s, 0);
		}

		// 부모(추천인) 초기화
		for (int i = 0; i < enroll.length; i++) {
			parent.put(enroll[i], referral[i]);
		}

		// 판매금액이 생길 시 수익을 update 하는 함수 호출
		for (int i = 0; i < seller.length; i++) {
			updateRevenue(seller[i], amount[i] * 100, revenue, parent);
		}


		// enroll의 등록 순서대로 수익을 정리
		int[] answer = new int[enroll.length];
		for(int i = 0; i < enroll.length; i++) {
			answer[i] = revenue.get(enroll[i]);
		}

		return answer;
	}


	private void updateRevenue(String child, int amount, Map<String, Integer> revenue, Map<String, String> parent) {
		int childRevenue = revenue.get(child); // 현재 금액 
		int passMoney = amount / 10; // 부모에게 줄 금액
		// 현재 금액 업데이트 
		revenue.put(child, childRevenue + amount - passMoney); 
		// 부모가 - 가 아니고, 부모에게 줘야할 돈이 있을 때 재귀
		if (!Objects.equals(parent.get(child), "-") && passMoney != 0) {    
			updateRevenue(parent.get(child), passMoney, revenue, parent);
		}
	}


	public static void main(String[] args) {
		Solution s = new Solution();
		System.out.println(Arrays.toString(
			s.solution(new String[] { "john", "mary", "edward", "sam", "emily", "jaimie", "tod", "young" }, new String[] { "-", "-", "mary", "edward", "mary", "mary", "jaimie", "edward" },
				new String[] { "young", "john", "tod", "emily", "mary" }, new int[] { 12, 4, 2, 5, 10 })));
	}
}

File: kwonmin/23_09/programmers/77486.py
# 칫솔 한 개당 이익은 100원으로 정해져 있음
# 즉 어마운트 * 100 하면 이익
# 이익을 타고 올라가는 과정
# 세그먼트 같은 느낌인데
# 느리게 갱신되는?
# 어? 이거 레이지 세그먼트인가? 흠?
# 그정도로 거창한 것 같지는 않고;;
# 이미 트리 구조니까
# 리프노드부터 올라오면서 큐에 단 한번씩만 계산하게 하는 게 맞을 듯?
# 결국 부모 노드가 누구인지는 알고 있으니까
# 부모 노드만 갱신하는 방식

# 이거 느리게 갱신하면 안되네
# 바로 갱신해야되네
# 한번에 하면, 그 1의 단위 없어질 때 어쩌구 조건 때문에 문제가 생길 듯

# 그럼 간단한 조작만 하면 될 듯?
# enroll이랑 referral 이랑 이거 하면서 인덱스 처리를 어떻게 할지.
# 결국 해당값에 맞는 enroll로 찾아가기 가 핵심
# 매핑하는 게 맞을듯.

from collections import defaultdict

def solution(enroll, referral, seller, amount):
    answer = [0]*len(enroll)
    idx = defaultdict(int)
    idx["-"] = -1
    for i in range(len(enroll)):
        idx[enroll[i]] = i

    # 추가로 뭘 안해도 찾아지지 않을까?

    def dfs(x,money):
        if x == -1:
            return
        
        dt = money // 10
        if dt:
            mine = money - dt
            answer[x] += mine
            dfs(idx[referral[x]], dt)
        else:
            answer[x] += money
    
    for j in range(len(seller)):
        dfs(idx[seller[j]], amount[j]*100)
    
    return answer


